core/distributed/vector.cpp

```diff
@@ -175,16 +177,65 @@ void Vector<ValueType>::read_distributed(
 
 
 template <typename ValueType>
-template <typename LocalIndexType, typename GlobalIndexType>
 void Vector<ValueType>::read_distributed(
-    const matrix_data<ValueType, GlobalIndexType>& data,
-    const Partition<LocalIndexType, GlobalIndexType>* partition)
+    const device_matrix_data<ValueType, int64>& data,
+    pointer_param<const Partition<int64, int64>> partition)
```

Comment:
like with gko::as(unique_ptr), templated functions can't be used here, because their parameters need to be an exact match, no implicit conversions allowed.